File: app-macos/Sources/RunnerForge/Models/DiskUsage.swift
import Foundation

/// Which side of the Cleanup page an item belongs to.
public enum DiskCategory: String, Sendable {
    /// Never deleted. Deleting it is what makes the next run slow.
    case keep
    /// Deleted on close and on demand.
    case purge
}

/// One row on the Cleanup page.
public struct DiskUsageItem: Sendable, Identifiable, Hashable {
    public let id = UUID()
    public let description: String
    public let category: DiskCategory
    public let bytes: Int64

    /// Why this item is kept, shown so KEEP does not look like clutter.
    public let reason: String?

    public var humanBytes: String { DiskUsageItem.format(bytes: bytes) }

    public init(description: String, category: DiskCategory, bytes: Int64, reason: String? = nil) {
        self.description = description
        self.category = category
        self.bytes = bytes
        self.reason = reason
    }

    public static func format(bytes: Int64) -> String {
        let units = ["B", "KB", "MB", "GB", "TB"]
        var value = Double(bytes)
        var unit = 0
        while value >= 1024, unit < units.count - 1 {
            value /= 1024
            unit += 1
        }
        return unit == 0 ? "\(bytes) B" : String(format: "%.1f %@", value, units[unit])
    }
}

/// The Cleanup page's whole model: two explicit lists and a total.
public struct DiskUsage: Sendable {
    public let keep: [DiskUsageItem]
    public let purge: [DiskUsageItem]

    public init(keep: [DiskUsageItem] = [], purge: [DiskUsageItem] = []) {
        self.keep = keep
        self.purge = purge
    }

    public var keepBytes: Int64 { keep.reduce(0) { $0 + $1.bytes } }
    public var reclaimableBytes: Int64 { purge.reduce(0) { $0 + $1.bytes } }

    public var humanKeepBytes: String { DiskUsageItem.format(bytes: keepBytes) }
    public var humanReclaimableBytes: String { DiskUsageItem.format(bytes: reclaimableBytes) }
}

File: app-windows/RunnerForge/Models/DiskUsage.cs
namespace RunnerForge.Models;

/// <summary>Which side of the Cleanup page an item belongs to.</summary>
public enum DiskCategory
{
    /// <summary>Never deleted. Deleting it is what makes the next run slow.</summary>
    Keep,

    /// <summary>Deleted on close and on demand.</summary>
    Purge,
}

/// <summary>One row on the Cleanup page.</summary>
public sealed record DiskUsageItem
{
    public required string Description { get; init; }
    public required DiskCategory Category { get; init; }
    public required long Bytes { get; init; }

    /// <summary>Why this item is kept, shown so KEEP does not look like clutter.</summary>
    public string? Reason { get; init; }

    public string HumanBytes => FormatBytes(Bytes);

    public static string FormatBytes(long bytes)
    {
        string[] units = ["B", "KB", "MB", "GB", "TB"];
        double value = bytes;
        int unit = 0;
        while (value >= 1024 && unit < units.Length - 1)
        {
            value /= 1024;
            unit++;
        }
        return unit == 0 ? $"{bytes} B" : $"{value:0.#} {units[unit]}";
    }
}

/// <summary>The Cleanup page's whole model: two explicit lists and a total.</summary>
public sealed record DiskUsage
{
    public IReadOnlyList<DiskUsageItem> Keep { get; init; } = [];
    public IReadOnlyList<DiskUsageItem> Purge { get; init; } = [];

    public long KeepBytes => Keep.Sum(i => i.Bytes);
    public long ReclaimableBytes => Purge.Sum(i => i.Bytes);

    public string HumanKeepBytes => DiskUsageItem.FormatBytes(KeepBytes);
    public string HumanReclaimableBytes => DiskUsageItem.FormatBytes(ReclaimableBytes);
}
